### ISSUE TITLE
move legacy modules into primary extension listing

### ISSUE THREAD

**Issue by tkellen** [tkellen/js-interpret#25]
In order to keep a single entry point in Liftoff for specifying module loaders per-extension, the extension listing should look something like this:
```js
{
  '.babel.js': 'babel/register',
  // ...
  '.ls': {
    main: 'coffee-script/register',
    legacy: ['coffee-script']
  },
  // ...
}
```

**Comment by tkellen** [comment]
This will address gulpjs/gulp#1071

**Comment by phated** [comment]
I think this is fine, but shouldn't everything be moved in there then? Like the install function, etc?

**Comment by tkellen** [comment]
Yeah, may as well.

**Comment by phated** [comment]
@tkellen does this make the liftoff implementation easier?

**Comment by tkellen** [comment]
Yes, massively. I don't know why I didn't think of this representation to begin with.

**Comment by tkellen** [comment]
Pretty sure this should do it:
```js
var extensions = {
  '.babel.js': {
    module: 'babel/register',
    register: function (module) {
      module({
        // register on .js extension due to https://github.com/joyent/node/blob/v0.12.0/lib/module.js#L353
        // which only captures the final extension (.babel.js -> .js)
        extensions: '.js'
      })
    }
  }
  '.cirru': 'cirru-script/lib/register',
  '.cjsx': 'node-cjsx/register',
  '.co': 'coco',
  '.coffee': ['coffee-script/register', 'coffee-script'],
  '.coffee.md': ['coffee-script/register', 'coffee-script'],
  '.csv': 'require-csv',
  '.iced': ['iced-coffee-script/register', 'iced-coffee-script'],
  '.iced.md': ['iced-coffee-script/register', 'iced-coffee-script'],
  '.ini': 'require-ini',
  '.js': null,
  '.json': null,
  '.json5': 'json5/lib/require',
  '.jsx': [
    {
      module: 'node-jsx',
      register: function (module) {
        module.install({
          extension: '.jsx',
          harmony: true
        });
      }
    }
  ],
  '.litcoffee': ['coffee-script/register', 'coffee-script'],
  '.liticed': ['iced-coffee-script/register', 'iced-coffee-script'],
  '.ls': ['LiveScript', 'livescript'],
  '.node': null,
  '.toml': 'toml-require',
  '.ts': ['typescript-register', 'typescript-require'],
  '.wisp': 'wisp/engine/node',
  '.xml': 'require-xml',
  '.yaml': 'require-yaml',
  '.yml': 'require-yaml'
};

var jsVariantExtensions = [
  '.js',
  '.babel.js',
  '.cirru',
  '.cjsx',
  '.co',
  '.coffee',
  '.coffee.md',
  '.iced',
  '.iced.md',
  '.jsx',
  '.litcoffee',
  '.liticed',
  '.ls',
  '.ts',
  '.wisp'
];

module.exports = {
  extensions: extensions,
  jsVariants: jsVariantExtensions.reduce(function (result, ext) {
    result[ext] = extensions[ext];
    return result;
  }, {})
};
```

If the value is a string, Liftoff will just try to require it. If the value is an object, it'll require the module and pass it into the register method. If the value is an array, Liftoff will iterate through them trying to find one that doesn't throw (the values of the array can be mixed, strings or objects).

...thoughts?

**Comment by phated** [comment]
@tkellen yeah, but `module` in an object needs to support an array also

**Comment by phated** [comment]
This would also allow us to support old versions of interpret in liftoff so it is a minor bump instead of major :+1:

**Comment by tkellen** [comment]
Do we have a use case for that now or are you just trying to keep things DRY in the event we need to use the same register method across multiple different possible modules?

**Comment by phated** [comment]
@tkellen updated my above comment.  I didn't read :stuck_out_tongue:

**Comment by tkellen** [comment]
Gotcha!

**Comment by phated** [comment]
@tkellen are you implementing or looking for someone to tackle?  I might have some time

**Comment by tkellen** [comment]
If you want to implement this in Liftoff, awesome! I think it'll be pretty straightforward.

**Comment by tkellen** [comment]
I'll cut a new major for interpret in a sec.

**Comment by phated** [comment]
@tkellen sounds good.  Make sure that the newest version of each module is in the first element of the array (e.g. `livescript` first instead of `LiveScript`)

**Comment by tkellen** [comment]
good catch. i'll do a PR for the version so you can sanity check

**Comment by tkellen** [comment]
resolved in 6131267d406f0b2b805ba71363daa2e9aa21e839
